## Part 1

In [173]:
import numpy as np
import pandas as pd
import re
from scipy.spatial.distance import cdist

In [174]:
data = [x for x in open("input.txt").read().splitlines()]

In [175]:
coord_str = [re.findall(r'\d+', entry) for entry in data]

In [176]:
coords = [[int(x) for x in lst] for lst in coord_str]

In [177]:
coord_arr = np.array(coords)


In [178]:
base_arr = np.zeros((coord_arr.max()+1,coord_arr.max()+1))


In [181]:
part2_arr

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [182]:
for i in range(0, len(coords)-1):
    x = coords[i][1]
    y = coords[i][0]
    base_arr[x][y] = i + 1


In [184]:
for i in range(0, len(base_arr)-1):
    for j in range(0, len(base_arr) - 1):
        base_coord = np.array([[i, j]])
        m_dist = cdist(base_coord, coords, 'cityblock') # manhattan distance from point to other points
        min_dist = np.where(m_dist == m_dist.min())
        if len(min_dist[0]) == 1:
            base_arr[i,j] = min_dist[1][0] +1

In [185]:
top_row = np.unique(base_arr[0])
last_row = np.unique(base_arr[-1])
first_col = np.unique(base_arr[:,0])
last_col = np.unique(base_arr[:,-1])
inf_list = list(np.unique(np.concatenate((top_row, last_row, first_col, last_col))))

In [186]:
area_tally = np.zeros(len(coords))
for i in range(0, len(base_arr) - 1):
    for j in range(0, len(base_arr) - 1):
        num = base_arr[i][j]
        if num not in inf_list:
            area_tally[int(num - 1)] += 1

In [187]:
max(area_tally)

3871.0

In [188]:
sum(area_tally)

60626.0

## Part 1 answer 3871

# Part 2

What is the size of the region containing all locations which have a total distance to all given coordinates of less than 10,000?

In [171]:
# for i in range(0, len(coords)-1):
#     x = coords[i][1]
#     y = coords[i][0]
#     part2_arr[x][y] = 1

In [192]:
part2_arr = np.zeros((coord_arr.max()+1,coord_arr.max()+1))

In [193]:
for i in range(0, len(part2_arr)-1):
    for j in range(0, len(part2_arr) - 1):
        base_coord = np.array([[i, j]])
        m_dist = cdist(base_coord, coords, 'cityblock') # manhattan distance from point to other points
        sum_dist = sum(m_dist[0])
        if sum_dist < 10000:
            part2_arr[i,j] += 1

In [196]:
len(part2_arr[part2_arr > 0])

44667

## Answer to Part 2 is 44,667